File: scripts/checkpoint_status.mjs
import { promises as fs } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const SCRIPT_DIR = path.dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = path.resolve(SCRIPT_DIR, '..');

function parseArgs(argv) {
  const out = {
    checkpointDir: path.resolve(REPO_ROOT, 'artifacts/checkpoints'),
    json: false,
  };

  for (const arg of argv) {
    if (arg.startsWith('--checkpoint-dir=')) {
      const value = arg.split('=')[1] ?? '';
      if (value.trim()) out.checkpointDir = path.resolve(process.cwd(), value.trim());
      continue;
    }
    if (arg === '--json') {
      out.json = true;
    }
  }

  return out;
}

function toModelKey(metrics) {
  const modelKey = metrics?.model_key;
  if (typeof modelKey === 'string' && modelKey.trim()) return modelKey;
  const provider = metrics?.provider ?? 'unknown';
  const modelId = metrics?.model_id ?? 'unknown';
  return `${provider}:${modelId}`;
}

function incCounter(map, key) {
  map.set(key, (map.get(key) ?? 0) + 1);
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const runsPath = path.resolve(args.checkpointDir, 'runs.jsonl');
  const statePath = path.resolve(args.checkpointDir, 'state.json');

  try {
    await fs.access(runsPath);
  } catch {
    if (args.json) {
      console.log(
        JSON.stringify(
          {
            checkpoint_dir: args.checkpointDir,
            runs_log: runsPath,
            state_file: statePath,
            exists: false,
            message: 'No checkpoint log found. Run the benchmark once first.',
          },
          null,
          2
        )
      );
      return;
    }
    console.log(`Checkpoint dir: ${args.checkpointDir}`);
    console.log('No checkpoint log found. Run the benchmark once first.');
    return;
  }

  const runsRaw = await fs.readFile(runsPath, 'utf8');
  const lines = runsRaw
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean);

  const latestByTask = new Map();

  for (const line of lines) {
    let parsed;
    try {
      parsed = JSON.parse(line);
    } catch {
      continue;
    }

    const taskKey = parsed?.task_key;
    const metrics = parsed?.metrics;
    if (!taskKey || !metrics) continue;
    latestByTask.set(taskKey, parsed);
  }

  let state = null;
  try {
    const raw = await fs.readFile(statePath, 'utf8');
    if (raw.trim()) {
      state = JSON.parse(raw);
    }
  } catch {
    // optional
  }

  let failed = 0;
  let successful = 0;
  const failedByModel = new Map();
  const successByModel = new Map();
  const failedErrorTypes = new Map();

  for (const record of latestByTask.values()) {
    const metrics = record.metrics;
    const modelKey = toModelKey(metrics);
    if (metrics?.error) {
      failed += 1;
      incCounter(failedByModel, modelKey);
      incCounter(failedErrorTypes, String(metrics.error));
    } else {
      successful += 1;
      incCounter(successByModel, modelKey);
    }
  }

  const summary = {
    checkpoint_dir: args.checkpointDir,
    runs_log: runsPath,
    state_file: statePath,
    raw_lines: lines.length,
    latest_task_records: latestByTask.size,
    successful_records: successful,
    failed_records: failed,
    failure_rate_pct: latestByTask.size > 0 ? (failed / latestByTask.size) * 100 : 0,
    state,
    by_model: Array.from(new Set([...successByModel.keys(), ...failedByModel.keys()]))
      .sort((a, b) => a.localeCompare(b))
      .map((model) => ({
        model_key: model,
        successful: successByModel.get(model) ?? 0,
        failed: failedByModel.get(model) ?? 0,
      })),
    top_errors: Array.from(failedErrorTypes.entries())
      .sort((a, b) => b[1] - a[1])
      .slice(0, 10)
      .map(([error, count]) => ({ count, error })),
  };

  if (args.json) {
    console.log(JSON.stringify(summary, null, 2));
    return;
  }

  console.log(`Checkpoint dir: ${summary.checkpoint_dir}`);
  console.log(`Records (latest by task): ${summary.latest_task_records}`);
  console.log(`Successful: ${summary.successful_records}`);
  console.log(`Failed: ${summary.failed_records} (${summary.failure_rate_pct.toFixed(2)}%)`);

  if (summary.state) {
    console.log(`Mode: ${summary.state.mode ?? 'unknown'} | Final: ${summary.state.final ?? false}`);
    console.log(`Updated at: ${summary.state.updated_at ?? 'n/a'}`);
  }

  console.log('\nBy model:');
  for (const row of summary.by_model) {
    console.log(`- ${row.model_key}: ok=${row.successful}, fail=${row.failed}`);
  }

  if (summary.top_errors.length > 0) {
    console.log('\nTop errors:');
    for (const row of summary.top_errors) {
      console.log(`- (${row.count}) ${row.error}`);
    }
  }
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: scripts/checkpoint_status.py
"""Summarize the state of a checkpoint dir. Python port of `scripts/checkpoint_status.mjs`.

Usage:
    uv run python scripts/checkpoint_status.py [--checkpoint-dir=PATH] [--json]
"""

from __future__ import annotations

import argparse
import json
import sys
from collections import Counter
from pathlib import Path
from typing import Any

from ocr_mini_bench.config.paths import PATHS


def _to_model_key(metrics: dict[str, Any]) -> str:
    model_key = metrics.get("model_key")
    if isinstance(model_key, str) and model_key.strip():
        return model_key
    provider = metrics.get("provider", "unknown")
    model_id = metrics.get("model_id", "unknown")
    return f"{provider}:{model_id}"


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(add_help=True)
    parser.add_argument(
        "--checkpoint-dir",
        type=Path,
        default=PATHS.checkpoint.root,
    )
    parser.add_argument("--json", action="store_true")
    args = parser.parse_args(argv)

    # Use absolute() not resolve() so we don't follow symlinks — matches
    # Node's path.resolve() behaviour (e.g. /tmp stays /tmp on macOS).
    checkpoint_dir: Path = args.checkpoint_dir.absolute()
    runs_path = checkpoint_dir / "runs.jsonl"
    state_path = checkpoint_dir / "state.json"

    if not runs_path.exists():
        if args.json:
            print(
                json.dumps(
                    {
                        "checkpoint_dir": str(checkpoint_dir),
                        "runs_log": str(runs_path),
                        "state_file": str(state_path),
                        "exists": False,
                        "message": "No checkpoint log found. Run the benchmark once first.",
                    },
                    indent=2,
                )
            )
            return 0
        print(f"Checkpoint dir: {checkpoint_dir}")
        print("No checkpoint log found. Run the benchmark once first.")
        return 0

    lines = [line for line in runs_path.read_text().splitlines() if line.strip()]
    latest_by_task: dict[str, dict[str, Any]] = {}
    for line in lines:
        try:
            parsed = json.loads(line)
        except json.JSONDecodeError:
            continue
        task_key = parsed.get("task_key")
        metrics = parsed.get("metrics")
        if not task_key or not metrics:
            continue
        latest_by_task[task_key] = parsed

    state: dict[str, Any] | None = None
    if state_path.exists():
        raw = state_path.read_text().strip()
        if raw:
            try:
                state = json.loads(raw)
            except json.JSONDecodeError:
                state = None

    failed = 0
    successful = 0
    failed_by_model: Counter[str] = Counter()
    success_by_model: Counter[str] = Counter()
    failed_error_types: Counter[str] = Counter()

    for record in latest_by_task.values():
        metrics = record["metrics"]
        model_key = _to_model_key(metrics)
        if metrics.get("error"):
            failed += 1
            failed_by_model[model_key] += 1
            failed_error_types[str(metrics["error"])] += 1
        else:
            successful += 1
            success_by_model[model_key] += 1

    by_model = [
        {
            "model_key": model,
            "successful": success_by_model.get(model, 0),
            "failed": failed_by_model.get(model, 0),
        }
        for model in sorted(set(success_by_model) | set(failed_by_model))
    ]
    top_errors = [
        {"count": count, "error": err}
        for err, count in failed_error_types.most_common(10)
    ]
    summary = {
        "checkpoint_dir": str(checkpoint_dir),
        "runs_log": str(runs_path),
        "state_file": str(state_path),
        "raw_lines": len(lines),
        "latest_task_records": len(latest_by_task),
        "successful_records": successful,
        "failed_records": failed,
        # Match TS JSON output: when the rate is exactly zero, JS serializes
        # it as `0` (int-shaped) rather than `0.0`. Mirror that here.
        "failure_rate_pct": (
            (failed / len(latest_by_task) * 100) if latest_by_task and failed else 0
        ),
        "state": state,
        "by_model": by_model,
        "top_errors": top_errors,
    }

    if args.json:
        print(json.dumps(summary, indent=2))
        return 0

    print(f"Checkpoint dir: {summary['checkpoint_dir']}")
    print(f"Records (latest by task): {summary['latest_task_records']}")
    print(f"Successful: {summary['successful_records']}")
    print(
        f"Failed: {summary['failed_records']} "
        f"({summary['failure_rate_pct']:.2f}%)"
    )

    if state:
        # JS-style lowercase booleans, to match TS output verbatim.
        final_str = str(state.get("final", False)).lower()
        print(f"Mode: {state.get('mode', 'unknown')} | Final: {final_str}")
        print(f"Updated at: {state.get('updated_at', 'n/a')}")

    print("\nBy model:")
    for row in by_model:
        print(f"- {row['model_key']}: ok={row['successful']}, fail={row['failed']}")

    if top_errors:
        print("\nTop errors:")
        for row in top_errors:
            print(f"- ({row['count']}) {row['error']}")

    return 0


if __name__ == "__main__":
    sys.exit(main())
